lib/streamlit/components/v1/component_registry.py

```diff
@@ -0,0 +1,125 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import inspect
+import os
+import threading
+from types import FrameType
+
+from streamlit.components.v1.base_component_registry import BaseComponentRegistry
+from streamlit.components.v1.custom_component import CustomComponent
+from streamlit.errors import CustomComponentError
+
+
+def _get_module_name(caller_frame: FrameType) -> str:
+    # Get the caller's module name. `__name__` gives us the module's
+    # fully-qualified name, which includes its package.
+    module = inspect.getmodule(caller_frame)
+    assert module is not None
+    module_name = module.__name__
+
+    # If the caller was the main module that was executed (that is, if the
+    # user executed `python my_component.py`), then this name will be
+    # "__main__" instead of the actual package name. In this case, we use
+    # the main module's filename, sans `.py` extension, as the component name.
+    if module_name == "__main__":
+        file_path = inspect.getfile(caller_frame)
+        filename = os.path.basename(file_path)
+        module_name, _ = os.path.splitext(filename)
+
+    return module_name
+
+
+def declare_component(
+    name: str,
+    path: str | None = None,
+    url: str | None = None,
+) -> CustomComponent:
+    """Create and register a custom component.
+
+    Parameters
+    ----------
+    name: str
+        A short, descriptive name for the component. Like, "slider".
+    path: str or None
+        The path to serve the component's frontend files from. Either
+        `path` or `url` must be specified, but not both.
+    url: str or None
+        The URL that the component is served from. Either `path` or `url`
+        must be specified, but not both.
+
+    Returns
+    -------
+    CustomComponent
+        A CustomComponent that can be called like a function.
+        Calling the component will create a new instance of the component
+        in the Streamlit app.
+
+    """
+
+    # Get our stack frame.
+    current_frame: FrameType | None = inspect.currentframe()
+    assert current_frame is not None
+    # Get the stack frame of our calling function.
+    caller_frame = current_frame.f_back
+    assert caller_frame is not None
+    module_name = _get_module_name(caller_frame)
+
+    # Build the component name.
+    component_name = f"{module_name}.{name}"
+
+    # Create our component object, and register it.
+    component = CustomComponent(
+        name=component_name, path=path, url=url, module_name=module_name
+    )
+    ComponentRegistry.instance().register_component(component)
+
+    return component
+
+
+class ComponentRegistry:
+    _instance_lock: threading.Lock = threading.Lock()
+    _instance: BaseComponentRegistry | None = None
```

Comment:
Hey @raethlein, one thing we have as an invariant, is that it should be possible to run the Stremlit app script as a raw Python script (e.g. `python app.py` instead of `streamlit run app.py`), this could be verified, but I think it is still the case. 

In that scenario, it makes sense to have a reasonable default value for `_instance` (just a dummy implementation that never records anything should be enough).

 In current behavior, we could fail when the server is not initialized.

Reply:
Ahh that's a great point! Setting the DefaultRegistry as the default sounds like a good idea

Reply:
I have changed it so that if `instance` is called before `initialize` was called, it instantiates the default registry